GetStream/stream-chat-android
# [CAS-1201] Add support for new pagination

https://stream-io.atlassian.net/browse/CAS-1201

### 🎯 Goal

Add a new way to paginate through search message results using limit and next/previous values.

### 🛠 Implementation details

Using request parameters directly instead of encapsulating them in `SearchMessagesRequest`. Motivation:

1. Easier to deprecate the old method. If `SearchMessagesRequest` class were used, then the old method and the new one would be conflicting as only difference would be the result payload.
1. Consistent with what we already have. Most of the methods in `ChatClient` already use individual parameters instead of wrapper class.
1. Easier to document each parameter.

### 🎨 UI Changes

![Screenshot 2021-09-01 at 03 18 49](https://user-images.githubusercontent.com/9600921/131592092-fb539e17-1430-4910-b6bd-e483d67bc06b.png)

### 🧪 Testing

1. Create an app with the new pagination enabled on the backend
2. Send 15 "Hi" messages and wait until they are properly indexed
3. Search for the messages containing "Hi" with the following query:
```kotlin
val firstPage = ChatClient.instance()
    .searchMessages(
        channelFilter = channelFilter,
        messageFilter = messageFilter,
        limit = 10,
        sort = QuerySort.desc("updated_at"),
    )
    .await()
```
4. Verify that the first chunk of messages has been loaded
5. Load the next page by using the following query:
```kotlin
ChatClient.instance()
    .searchMessages(
        channelFilter = channelFilter,
        messageFilter = messageFilter,
        limit = 10,
        next = firstPage.data().next, // the next page from previous query
    )
    .await()
```
6. Verify that the next 5 messages were loaded

### ☑️ Checklist


### 🎉 GIF

![giphy (8)](https://user-images.githubusercontent.com/9600921/131589551-2f8f4403-9a6e-48fc-b26d-d3e719c486b8.gif)
